### ISSUE TITLE
compact declaration of internal set & ptr

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#245]
**Reported by @stefanb on 23 Jul 2003 09:50 UTC**
Current set declaration looks like:
field = set
  field->subfield=char[
  field->otherSubfield=date ;The date 
  field->choice=char{"a"=1, "b"=2}
  field->someMore=text

first i thought of omitting the field name before "->":
field = set
  ->subfield=char[33](33]) 
  ->otherSubfield=date ;The date 
  ->choice=char{"a"=1, "b"=2}
  ->someMore=text

then i realized we were already talking about it and found it written down in 
the spec at 
http://www.makumba.org/makumba-spec.html#Structural (relational) types
but is nowhere in bugzilla yet. 
It talks about even better style:
field = set {
  subfield=char[33] 
  otherSubfield=date ;The date 
  choice=char{"a"=1, "b"=2}
  someMore=text
}

Migrated-From: http://trac.makumba.org/ticket/517

**Comment by ghost** [comment]
**Comment by @stefanb on 13 Oct 2003 21:22 UTC**
*** Bug 607 has been marked as a duplicate of this bug. ***

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 22 Jan 2010 14:48 UTC**
having re-implemented the MDD parser, I am not very much in favor of this notation, as it is rather complex to parse in the new parser: it mixes the usage of brackets and of line breaks.

as brackets are now being used for int enum, function bodies and validation rule bodies, it seems to me that using them for describing collections would be rather confusing, not only for the parser, but also for the user.

if there are no strong opinions, I will thus close the entry a week from now

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 22 Jan 2010 18:03 UTC**
(In reply to comment #2)
> having re-implemented the MDD parser, I am not very much in favor of this
> notation, as it is rather complex to parse in the new parser: it mixes the
> usage of brackets and of line breaks.


I don't know much about that, unfortunately.
still, a question - can ANTLR proceed only line per line? 

> as brackets are now being used for int enum, function bodies and validation
> rule bodies, it seems to me that using them for describing collections would be
> rather confusing, not only for the parser, but also for the user.


I don't think it is confusing for the user - in all those cases, the brackets are used to define the body of the field/function. the same would be here with the set.

> if there are no strong opinions, I will thus close the entry a week from now


I think that such a notation would be much more readable, more intuitive, and less error-prone for the user. Thus, if it is reasonably possible, I would not discard the idea that easily.

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 26 Jan 2010 13:19 UTC**
(In reply to comment #3)
> I don't know much about that, unfortunately.
> still, a question - can ANTLR proceed only line per line? 

no, it processes everything. however given that the original MDD parser was a line-by-line parser, the new MDD grammar is also line-by-line, i.e. the root rules have to know about line-breaks.

the main problem in the proposed design is that it mixes brackets and line-breaks, in essence, it can only work with a line-by-line parser, but one that processes lines one after another (like the old recordparser)

something that would perhaps be easier to deal with is to have a separator :

field = set {
   subfield=char[  otherSubfield=date,
   choice=char{"a"=1, "b"=2},
   someMore=text
}


> I think that such a notation would be much more readable, more intuitive, and
> less error-prone for the user. Thus, if it is reasonably possible, I would not
> discard the idea that easily.

sure, we can keep the idea, but I will most probably not be the one to implement it, as this means re-implementing most of the MDD parser:

everything in { } is treated as a literal and passed to another grammar that parses its contents (or to the function inliner), because this was the only way to deal with the functions or regexp validation rules, that can contain virtually anything. and until now I cannot think of another way of dealing with this (regular expressions and function definitions in the MDD). so, if subfields were to be inside of { } this means that the mdd-expression grammar would need to know about field, types, subfields, validation rules, ... and this is extremely clumsy.

in conclusion, we should not design more stuff that goes into { }, or use something else than { }, e.g. [ ](33],
) (but IMO that looks strange to say the least)

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:22 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**
